dynamodb/schema.go

```diff
@@ -0,0 +1,358 @@
+	typeNumberFloatSet = "NumberFloatSet"
+	typeBinarySet      = "BinarySet"
+
+	errThreshold      = float64(0.001)
+	conflictThreshold = float64(0.05)
+)
+
+type dynamoClient interface {
+	ListTables(input *dynamodb.ListTablesInput) (*dynamodb.ListTablesOutput, error)
+	DescribeTable(input *dynamodb.DescribeTableInput) (*dynamodb.DescribeTableOutput, error)
+	Scan(input *dynamodb.ScanInput) (*dynamodb.ScanOutput, error)
+}
+
+// ProcessSchema performs schema conversion for source tables in a DynamoDB
+// database. We use the standard APIs to obtain source table's schema
+// information, including primary keys and secondary indexes. DynamoDB is a
+// schemaless database that some optional attributes can be missed or has
+// different data types in some rows. Therefore, we have to sample a number of
+// rows to infer the schema for optional attributes.
+func ProcessSchema(conv *internal.Conv, client dynamoClient, tables []string, sampleSize int64) error {
+	if len(tables) == 0 {
+		var err error
+		tables, err = listTables(client)
+		if err != nil {
+			return err
+		}
+		if len(tables) == 0 {
+			return fmt.Errorf("no DynamoDB table exists under this account")
+		}
+	}
+	for _, t := range tables {
+		if err := processTable(conv, client, t, sampleSize); err != nil {
+			return err
+		}
+	}
+	schemaToDDL(conv)
+	conv.AddPrimaryKeys()
+	return nil
+}
+
+func listTables(client dynamoClient) ([]string, error) {
+	var tables []string
+	var lastEvaluatedTableName *string
+
+	for {
+		input := &dynamodb.ListTablesInput{}
+		if lastEvaluatedTableName != nil {
+			input.ExclusiveStartTableName = lastEvaluatedTableName
+		}
+
+		result, err := client.ListTables(input)
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range result.TableNames {
+			tables = append(tables, *t)
+		}
+
+		if result.LastEvaluatedTableName == nil {
+			return tables, nil
+		}
+		lastEvaluatedTableName = result.LastEvaluatedTableName
+	}
+}
+
+func processTable(conv *internal.Conv, client dynamoClient, table string, sampleSize int64) error {
+	dySchema := dynamoDBSchema{TableName: table}
+	err := dySchema.parseIndexes(client)
+	if err != nil {
+		return err
+	}
+	stats, count, err := dySchema.scanSampleData(client, sampleSize)
+	if err != nil {
+		return err
+	}
+	dySchema.inferDataTypes(stats, count)
+	conv.SrcSchema[table] = dySchema.genericSchema()
+	return nil
+}
+
+type dynamoDBSchema struct {
+	TableName   string
+	ColumnNames []string
+	ColumnTypes map[string]colType
+	PrimaryKeys []string
+	SecIndexes  []index
+}
+
+type colType struct {
+	Name     string
+	Nullable bool
+}
+
+type index struct {
+	Name string
+	Keys []string
+}
+
+func (s *dynamoDBSchema) parseIndexes(client dynamoClient) error {
+	input := &dynamodb.DescribeTableInput{
+		TableName: aws.String(s.TableName),
+	}
+
+	result, err := client.DescribeTable(input)
+	if err != nil {
+		return fmt.Errorf("failed to make a DescribeTable API call for table %v: %v", s.TableName, err)
+	}
+
```

Comment:
Should we use AttributeDefinitions to get types for keys?

Also, if you give attribute definitions for other columns, how are they treated? Are the specified types enforced? (my guess is probably not, but I don't know for sure).
Should we use types for these addition attribute definitions to help build the schema?

Reply:
The `AttributeDefinitions` only returns the columns of a primary key, not any optional columns. 

When I try to create a table via: 

```bash
aws dynamodb create-table --table-name Migration \                                                                
                                  --attribute-definitions AttributeName=Username,AttributeType=S AttributeName=Unused,AttributeType=S \
                                  --key-schema AttributeName=Username,KeyType=HASH \
                                  --provisioned-throughput ReadCapacityUnits=5,WriteCapacityUnits=5
```

As you can see, `attribute-definitions` does not match `key-schema` and then I got the following error: 

```bash
An error occurred (ValidationException) when calling the CreateTable operation: One or more parameter values were invalid: Number of attributes in KeySchema does not exactly match number of attributes defined in AttributeDefinitions. 
```

What I mean is that `AttributeDefinitions` must match `KeySchema`, which is the primary key. No optional columns will appear in `AttributeDefinitions`. 

> Should we use AttributeDefinitions to get types for keys?

Yes, we can, but only for columns of a primary key. We can use the `AttributeDefinitions` as a base schema, which only contains the types of columns in a primary key. Then, use sampling to guess the types of optional columns. 

The current implementation didn't do this because the types of columns in a primary key are forced and they have an absolute win in the sampling process. I try to avoid the process to add optional columns to a base table. If you think using `AttributeDefinitions` as a base table is a better idea, then I can update it. 

BTW: worth to mention that for secondary indexes, there are no `AttributeDefinitions`: 

Local secondary index: 

```json
"LocalSecondaryIndexes": [
            {
                "IndexName": "AlbumTitleIndex",
                "KeySchema": [
                    {
                        "AttributeName": "Artist",
                        "KeyType": "HASH"
                    },
                    {
                        "AttributeName": "AlbumTitle",
                        "KeyType": "RANGE"
                    }
                ],
                "Projection": {
                    "ProjectionType": "INCLUDE",
                    "NonKeyAttributes": [
                        "Genre",
                        "Year"
                    ]
                },
                "IndexSizeBytes": 0,
                "ItemCount": 0,
                "IndexArn": "arn:aws:dynamodb:us-west-2:123456789012:table/MusicCollection/index/AlbumTitleIndex"
            }
        ]
```

Global secondary index: 

```json
"GlobalSecondaryIndexes": [
            {
                "IndexName": "GameTitleIndex",
                "KeySchema": [
                    {
                        "AttributeName": "GameTitle",
                        "KeyType": "HASH"
                    },
                    {
                        "AttributeName": "TopScore",
                        "KeyType": "RANGE"
                    }
                ],
                "Projection": {
                    "ProjectionType": "INCLUDE",
                    "NonKeyAttributes": [
                        "UserId"
                    ]
                },
                "IndexStatus": "CREATING",
                "ProvisionedThroughput": {
                    "NumberOfDecreasesToday": 0,
                    "ReadCapacityUnits": 10,
                    "WriteCapacityUnits": 5
                },
                "IndexSizeBytes": 0,
                "ItemCount": 0,
                "IndexArn": "arn:aws:dynamodb:us-west-2:123456789012:table/GameScores/index/GameTitleIndex"
            }
        ]
```

Ref: https://docs.aws.amazon.com/cli/latest/reference/dynamodb/create-table.html

Reply:
Ok, got it.
Thanks for all of the clarifications.
Yes, that makes perfect sense now.
And agreed that if key type is enforced, then the sampled data is guaranteed to give us the right type.

I guess the only corner case is: what happens if the table is empty? Hmm, thinking about it, I guess all you could do in that case is use the meta data to add a Spanner table with just the key column and no other columns, which isn't very useful. Ok, let's go with what you have.